### PR TITLE
Add cluster auto-scaler documentation

### PR DESCRIPTION
**Pull request by kalexand-rh**
Documents how to setup, deploy and validate the cluster auto-scaler is
functioning correctly. Only for AWS.

Follow up to https://github.com/openshift/openshift-docs/pull/11924 for edits.

### CODE REVIEW

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (kalexand-rh) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: @sunzhaohua2, is this what you meant in https://github.com/openshift/openshift-docs/pull/11924#issuecomment-422304897?

**Comment by kalexand-rh** [comment]
@frobware, will you PTAL?

@sunzhaohua2, @mdshuai, will you please approve the changes so I can merge? 

@openshift/team-documentation, will you please review?

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (sunzhaohua2) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: I think this step can be moved to the part of "Prerequisites" as a note. It is about deploy the cluster need to pay attention to not deploy cluster-autoscaler.

Comment (frobware) [reply admin_guide/topics/deploying-cluster-auto-scaler.adoc]: Yes - to confirm, the `bootstrap-autoapprover` needs to be running on the _existing_ cluster.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Depending on what {product-title} is here, is `an` always correct?

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: I don't know how we spell these examples out in the existing docs, but when specifying paths I tend to go with: `'/path/to/inventory'`, the presence of the `'/'` enforces that it is a path.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: nit; the same host doesn't matter. You could actually run these on commands on the cluster itself which, given the nature of the examples, is what I do when testing.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: It's not "for the namespace". It's for the autoscaler.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: It's not for the namespace, it's for the cluster-autoscaler.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: ... AWS credentials so that the ASG has permissions to launch new instances.

**Review by frobware — APPROVED** [review]
Some nits and corrections, but overall /lgtm.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: For the collection's this file is in, {product-title} renders to either OpenShift Container Platform or OKD, so yes. :)

File: admin_guide/topics/creating-golden-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-golden-images-cluster-auto-scaler.adoc]: Should be `openshift_deployment_type=origin`

File: admin_guide/topics/creating-golden-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-golden-images-cluster-auto-scaler.adoc]: Add trailing `\`

File: admin_guide/topics/creating-golden-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-golden-images-cluster-auto-scaler.adoc]: Add trailing `\`

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: s/to run the/for the

File: admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc]: s/which you use.../which you use when deploying the autoscaler deployment

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: For consistency: `your-aws-secret-access-key` - note the last hypen.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: Add leading `$`

File: admin_guide/topics/testing-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/testing-auto-scaler.adoc]: I think we should be less precise. depends on region, image type, the prevailing wind and probably many other things.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: We need to mention that you may need to re-run the AWS playbooks on the existing cluster to deploy the bootstrap auto approver components if this component is `False`.

Comment (frobware) [reply admin_guide/topics/deploying-cluster-auto-scaler.adoc]: The more I think about this is... as listed as prerequisite it should say, your existing cluster needs the bootstrap autoapprover components to be running. You can check this with:

```console
$ [ec2-user@ip-172-18-9-89 ~]$ oc get pods --all-namespaces | grep bootstrap-autoapprover
openshift-infra         bootstrap-autoapprover-0                         1/1       Running   0          5h
```

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: s/uses this secret during deployments/uses this secret to launch instances within AWS

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Should be `openshift_deployment_type=openshift-enterprise`

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Should be `openshift_deployment_type=origin`

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/deploying-cluster-auto-scaler.adoc]: That's fair, but I think we need to move it out of the prereqs and into the steps.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (sunzhaohua2) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Should be "openshift_deployment_type: openshift-enterprise"

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (sunzhaohua2) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Should be "openshift_deployment_type: origin"

**Comment by frobware** [comment]
Zhaohua Sun <notifications@github.com> writes:

> sunzhaohua2 commented on this pull request.
>
>
>
>> +ansible_ssh_user=ec2-user
> +openshift_clusterid=mycluster
> +ansible_become=yes
> +
> +[masters]
> +[etcd]
> +[nodes]
> +----
> +
> +. Create provisioning file, *_build-ami-provisioning-vars.yaml_*, on your local
> +host:
> ++
> +[source,yaml]
> +----
> +ifdef::openshift-enterprise[]
> +openshift_deployment_type=openshift-enterprise
>
> Should be "openshift_deployment_type: openshift-enterprise"

Yes.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Correct. Not sure what I was thinking yesterday.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Yep.

File: install/configuring_inventory_file.adoc
Comment (kalexand-rh) [inline install/configuring_inventory_file.adoc]: @sdodson, @frobware,  I added the parameter to the General Cluster Variables table. Please let me know if you see more changes.

**Comment by kalexand-rh** [comment]
@sunzhaohua2, @frobware, if there are no more changes, will you both agree that this PR looks good so I can merge?

**Comment by frobware** [comment]
> if there are no more changes, will you both agree that this PR looks good so I can merge?

I have no other changes. Thanks!

File: admin_guide/topics/about-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/about-cluster-auto-scaler.adoc]: s/an ASGs/an ASG

File: admin_guide/topics/about-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/about-cluster-auto-scaler.adoc]: @kalexand-rh Did you check the AWS docs for this abbreviation? I took a quick look and didn't see it. It looks a bit odd that _groups_ is small G when spelled out; but becomes uppercase in the abbreviation. Doesn't affect understanding. Just made me question if the name is _Auto Scaling Group_ (which it is not).

File: admin_guide/topics/about-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/about-cluster-auto-scaler.adoc]: s/A LC/An LC

File: admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc]: s/by using/using ??

File: admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc]: @kalexand-rh 
s/by using/using ??

File: admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc]: IBM Style says:

using: To avoid ambiguity, replace this gerund with either “by using” or “that uses.”

File: admin_guide/topics/about-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/about-cluster-auto-scaler.adoc]: Good point. I can't find ASG in the docs, but they do use Auto Scaling group :https://docs.aws.amazon.com/autoscaling/ec2/userguide/create-asg-launch-template.html

File: admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc]: @kalexand-rh 
I know we try to avoid _will_, but I think _which you will use when..._ sound clearer to me.

s/deploy the auto-scaler deployment/deploy the auto-scaler components

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: @kalexand-rh 
What does _seal_ mean here? Would a typical admin know?

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: @kalexand-rh 
Did we run `deploy_cluster.yml` previously?

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: @kalexand-rh 
Should this output have headings, as in:
`NAME                                READY             STATUS      RESTARTS          AGE`

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/deploying-cluster-auto-scaler.adoc]: @kalexand-rh 
Is _cluster-reader policy_ the right term? I searched in Atom and see _`cluster-reader` permission_ several times, but not _policy_.

File: admin_guide/topics/testing-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/testing-auto-scaler.adoc]: @kalexand-rh 
s/works../works.

File: admin_guide/topics/testing-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/testing-auto-scaler.adoc]: @kalexand-rh 
_ without first increasing the number of compute nodes_ which is done automatically by the auto scaler?

File: admin_guide/topics/testing-auto-scaler.adoc
Comment (mburke5678) [inline admin_guide/topics/testing-auto-scaler.adoc]: @kalexand-rh 
_These pending pods have nowhere to run_ sounds a little too cute. 
How about combining the first two sentenced_The pods cannot be scheduled until the auto scaler provisions new nodes_
(Poor pods, nowhere to run....)

File: install/configuring_inventory_file.adoc
Comment (mburke5678) [inline install/configuring_inventory_file.adoc]: s/if you will enable/you enable ?

**Comment by mburke5678** [comment]
@kalexand-rh A few nits a few questions. Otherwise LGreatTM

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: @frobware, what's a plainer way to say "seal" here?

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/deploying-cluster-auto-scaler.adoc]: @mburke5678, you do when you install the cluster, which is one of the prereqs.

File: install/configuring_inventory_file.adoc
Comment (kalexand-rh) [reply install/configuring_inventory_file.adoc]: Normally, yes, but I'm afraid that "you enable" might imply that you can enable the auto-scaler **now**, which you can't.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (mburke5678) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: @kalexand-rh I am not sure what that means.

Comment (frobware) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: > what's a plainer way to say "seal" here?

Maybe just drop that phrase? All we care about is creating the image.

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [reply admin_guide/topics/deploying-cluster-auto-scaler.adoc]: Here in case you want them: 

```
NAMESPACE               NAME                                             READY     STATUS    RESTARTS   AGE
```

File: admin_guide/topics/about-cluster-auto-scaler.adoc
Comment (frobware) [reply admin_guide/topics/about-cluster-auto-scaler.adoc]: > s/A LC/An LC

Seems harder (to say out loud at least). "An launch configuration" vis-a-vis "A launch configuration".

Comment (mburke5678) [reply admin_guide/topics/about-cluster-auto-scaler.adoc]: @frobware For acronyms we use a or an based how the acronym sounds. An ell-cee (LC)

Comment (frobware) [reply admin_guide/topics/about-cluster-auto-scaler.adoc]: ic

File: admin_guide/topics/deploying-cluster-auto-scaler.adoc
Comment (frobware) [reply admin_guide/topics/deploying-cluster-auto-scaler.adoc]: Probably better as cluster-reader role.

File: admin_guide/topics/about-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/about-cluster-auto-scaler.adoc]: I don't see ASG or LC in the AWS docs, so I'm expanding all the references.

File: admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/creating-LC-and-ASG-cluster-auto-scaler.adoc]: I think this one's close enough to when you're doing the thing to not need will.

And it's really a deployment. :(

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: This is actually wrong. I was right before, and wrong to change it so many times. Sorry!

Inventory entries are separated by ':'. The extra provisioning files are YAML and are separated by '='.

This entry should be: `openshift_deployment_type=openshift-enterprise`.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: This entry should be: `openshift_deployment_type=origin`.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: This line is missing the trailing backslash `\`.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (frobware) [inline admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: This line is missing the trailing backslash `\`.

File: admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc
Comment (kalexand-rh) [reply admin_guide/topics/creating-primed-images-cluster-auto-scaler.adoc]: Ah! Thank you for explaining the difference between the two references! I own my part in this - I was resetting them both to be the same. :/

**Comment by frobware** [comment]
With these changes https://github.com/openshift/openshift-docs/pull/12181/commits/0845a280ce9203836c7face4a7775d1cbd0c2908 I went through the docs again today and deployed the autoscaler against a deployment that didn't have it present and deployed and the docs worked for me.

**Comment by kalexand-rh** [comment]
@vikram-redhat, we addressed the last of the concerns that QE brought up in comments. @sdodson said that it looked fine on the original PR. Since @frobware got the instructions to work and QE's gone this week, I'm going to merge it. Please let me know if you want me to revert the change pending a LGTM from QE.

**Pull request merged**
